src/handler.rs

```diff
@@ -3,14 +3,78 @@ use std::future::Future;
 use std::pin::Pin;
 use std::sync::Arc;
 
-use futures::channel::oneshot::Sender as SyncSender;
+use futures::channel::oneshot::{self, Sender as SyncSender};
+use futures::FutureExt;
 
 use crate::actor::{Actor, AsyncContext};
 use crate::address::Addr;
 use crate::context::Context;
-use crate::fut::ActorFuture;
+use crate::fut::{wrap_future, ActorFuture};
 use crate::WrapFuture;
 
+pub struct TempRef<T> {
+    inner: *mut T,
+}
+
+unsafe impl<T: Send> Send for TempRef<T> {}
+
+impl<T> TempRef<T> {
+    pub fn as_ref(&self) -> &T {
+        unsafe {
+            let x = self.inner;
+            &*x
+        }
+    }
+    pub fn as_mut(&mut self) -> &mut T {
+        unsafe {
+            let x = self.inner;
+            &mut *x
+        }
+    }
+}
+
+pub trait AsyncHandler<M>
+where
+    Self: Actor,
+    <Self as Actor>::Context: AsyncContext<Self>,
+    M: AsyncMessage,
+{
+    type Result: Future<Output = M::Result> + Sized + 'static;
+
+    fn handle(
+        actor: TempRef<Self>,
+        msg: M,
+        ctx: TempRef<Self::Context>,
+    ) -> Self::Result;
+}
+
+impl<M, I: Send + 'static, AH> Handler<M> for AH
+where
+    AH: AsyncHandler<M> + 'static,
+    <AH as Actor>::Context: AsyncContext<AH>,
+    M: AsyncMessage<Result = I> + 'static,
+{
+    type Result = ResponseFuture<AsyncResponse<M>>;
+
+    fn handle(&mut self, msg: M, ctx: &mut Self::Context) -> Self::Result {
+        let actor = TempRef { inner: self };
+
+        let tmp_ctx = TempRef { inner: ctx };
```

Comment:
As I understand, with actor we can be sure that it is safe to have a mutable reference if we don't call `handle()` on it.

With context I'm not sure. Especially because we are using it below to wait for the future result.

Maybe, wrapper for the context can be something very different, that has limited of different interface, just allowing to get the address, for example.